xcp_d/utils/concatenation.py

```diff
@@ -0,0 +1,452 @@
+from xcp_d.utils.qcmetrics import compute_dvars
+from xcp_d.utils.utils import get_segfile
+from xcp_d.utils.write_save import read_ndata
+
+_pybids_spec = loads(Path(_pkgres("xcp_d", "data/nipreps.json")).read_text())
+path_patterns = _pybids_spec["default_path_patterns"]
+LOGGER = logging.getLogger("nipype.interface")
+
+
+def concatenate_derivatives(fmridir, outputdir, work_dir, subjects, cifti):
+    """Concatenate derivatives.
+
+    This function does a lot more than concatenate derivatives.
+    It also makes DCAN QC files, creates figures, and copies figures from the preprocessed
+    dataset to the post-processed dataset.
+
+    TODO: Move concatenation to *inside* main workflow.
+    That way we can feed files in directly instead of searching for them,
+    and we can pass the already-initialized fMRIPrep BIDSLayout.
+
+    Parameters
+    ----------
+    fmridir : str
+        Path to preprocessed derivatives (not xcpd post-processed derivatives).
+    outputdir : str
+        Path to location of xcpd derivatives.
+    work_dir : str
+        Working directory.
+    subjects : list of str
+        List of subjects to run concatenation on.
+    cifti : bool
+        Whether xcpd was run on CIFTI files or not.
+    """
+    # NOTE: The config has no effect when derivatives is True :(
+    # At least for pybids ~0.15.1
+    layout = BIDSLayout(outputdir, validate=False, derivatives=True)
+    layout_fmriprep = BIDSLayout(fmridir, validate=False, derivatives=True)
+
+    if cifti:
+        tsv_extensions = [".ptseries.nii"]
+        img_extensions = [".dtseries.nii"]
+    else:
+        tsv_extensions = [".tsv"]
+        img_extensions = [".nii.gz"]
+
+    for subject in subjects:
+        if subject.startswith("sub-"):
+            subject = subject[4:]
+
+        sessions = layout.get_sessions(subject=subject)
+        if not sessions:
+            sessions = [None]
+
+        for session in sessions:
+            base_entities = {
+                "subject": subject,
+                "session": session,
+                "datatype": "func",
+            }
+            tasks = layout.get_tasks(
+                desc="denoised",
+                suffix="bold",
+                extension=img_extensions,
+                **base_entities,
+            )
+            for task in tasks:
+                task_entities = base_entities.copy()
+                task_entities["task"] = task
+
+                motion_files = layout.get(
+                    desc=["filtered", None],
+                    suffix="motion",
+                    extension=".tsv",
+                    **task_entities,
+                )
+                if len(motion_files) == 0:
+                    continue
+                elif len(motion_files) == 1:
+                    # Make DCAN HDF5 file from single motion file
+                    dcan_df_file = (
+                        f"{'.'.join(motion_files[0].split('.')[:-1])}-DCAN.hdf5"
```

Comment:
```suggestion
                        f"{'.'.join(motion_files[0].path.split('.')[:-1])}-DCAN.hdf5"
```

I got the error below on tasks with a single run; with the above change it completed w/o error and output looks good.  

My test data was the colornest CIFTI w/ FreeSurfer, with the second of the two "task-rest" runs removed (including the corresponding files in the fMRIPrep figures dir).
```
Traceback (most recent call last):
  File "/usr/local/miniconda/bin/xcp_d", line 8, in <module>
    sys.exit(main())
  File "/usr/local/miniconda/lib/python3.8/site-packages/xcp_d/cli/run.py", line 648, in main
    failed_reports = generate_reports(
  File "/usr/local/miniconda/lib/python3.8/site-packages/xcp_d/interfaces/report_core.py", line 153, in generate_reports
    concatenate_derivatives(
  File "/usr/local/miniconda/lib/python3.8/site-packages/xcp_d/utils/concatenation.py", line 99, in concatenate_derivatives
    f"{'.'.join(motion_files[0].split('.')[:-1])}-DCAN.hdf5"
  File "/usr/local/miniconda/lib/python3.8/site-packages/bids/layout/models.py", line 224, in __getattr__
    raise AttributeError("%s object has no attribute named %r" %
AttributeError: BIDSDataFile object has no attribute named 'split'
```

Reply:
Great catch! Thanks!